substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,420 @@
+                            contract with staking.",
+					);
+				},
+				StakerStatus::Idle => (), // nothing to see here.
+			}
+		}
+	}
+
+	// Fired when someone sets their intention to nominate.
+	//
+	// Note: it is assumed that who's staking state is updated *before* this method is called.
+	fn on_nominator_add(who: &T::AccountId) {
+		let nominator_vote = Self::weight_of(Self::active_vote_of(who));
+
+		let _ = T::VoterList::on_insert(who.clone(), nominator_vote).defensive_proof(
+			"staker should not exist in VoterList, as per the contract with staking.",
+		);
+
+		// If who is a nominator, update the vote weight of the nominations if they exist. Note:
+		// this will update the score of up to `T::MaxNominations` validators.
+		match T::Staking::status(who).defensive() {
+			Ok(StakerStatus::Nominator(nominations)) =>
+				for t in nominations {
+					Self::update_score::<T::TargetList>(
+						&t,
+						StakeImbalance::Positive(nominator_vote.into()),
+					)
+				},
+			Ok(StakerStatus::Idle) | Ok(StakerStatus::Validator) | Err(_) => (), // nada.
+		};
+
+		log!(debug, "on_nominator_add: {:?}. role: {:?}", who, T::Staking::status(who),);
+	}
+
+	// Fired when someone sets their intention to validate.
+	//
+	// A validator is also considered a voter with self-vote and should be added to
+	// [`Config::VoterList`].
+	//
+	// Note: it is assumed that who's staking state is updated *before* calling this method.
+	fn on_validator_add(who: &T::AccountId) {
+		let _ = T::TargetList::on_insert(who.clone(), Self::active_vote_of(who)).defensive_proof(
+			"staker should not exist in TargetList, as per the contract with staking.",
+		);
+
+		log!(debug, "on_validator_add: {:?}. role: {:?}", who, T::Staking::status(who),);
+
+		// a validator is also a nominator.
+		Self::on_nominator_add(who)
+	}
+
+	// Fired when someone removes their intention to nominate, either due to chill or validating.
+	//
+	// Note: it is assumed that who's staking state is updated *before* the caller calling into
+	// this method. Thus, the nominations before the nominator has been removed from staking are
+	// passed in, so that the target list can be updated accordingly.
+	fn on_nominator_remove(who: &T::AccountId, nominations: Vec<T::AccountId>) {
+		let nominator_vote = Self::weight_of(Self::active_vote_of(who));
+
+		log!(
+			debug,
+			"on_nominator_remove: {:?} with {:?}. impacting {:?}",
+			who,
+			nominator_vote,
+			nominations,
+		);
+
+		// updates the nominated target's score. Note: this may update the score of up to
+		// `T::MaxNominations` validators.
+		for t in nominations.iter() {
+			Self::update_score::<T::TargetList>(&t, StakeImbalance::Negative(nominator_vote.into()))
+		}
+
+		let _ = T::VoterList::on_remove(&who).defensive_proof(
+			"the nominator exists in the list as per the contract with staking; qed.",
+		);
+	}
+
+	// Fired when someone removes their intention to validate, either due to chill or nominating.
```

Comment:
Would be nice if there's some docs on 'chilling' and other terms used here to deep link to

```suggestion
	/// Fired when someone removes their intention to validate, either due to chill or nominating.
```